### ISSUE TITLE
Add UnderScoreCase support for System.Text.Json

### ISSUE THREAD

**Issue by hez2010** [dotnet/corefx#39564]
Can you add UnderScoreCase in System.Text.Json?
This naming policy is also common used.

**Comment by Gnbrkm41** [comment]
If you have any designs in mind, could you update the top post with what the proposed APIs would look like / how commonly used it is / etc? Here's what I think is a good example: https://github.com/dotnet/corefx/issues/28933#issue-312415055. This should help people better understand the problem/importance of this issue, and may come up with improvements.

**Comment by khellang** [comment]
## API Proposal

```diff
namespace System.Text.Json
{
    public abstract class JsonNamingPolicy
    {
        protected JsonNamingPolicy() { }
        public static JsonNamingPolicy CamelCase { get { throw null; } }
+       public static JsonNamingPolicy SnakeCase { get { throw null; } }
        internal static JsonNamingPolicy Default { get { throw null; } }
        public abstract string ConvertName(string name);
    }

+   internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+   {
+       public override string ConvertName(string name) { throw null; }
+   }
}
```

## Behavior Proposal

I propose the same behavior as Newtonsoft.Json, just like the existing camel case behavior. The implementation is [here](https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json/Utilities/StringUtils.cs#L200-L276) and the tests are [here](https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json.Tests/Utilities/StringUtilsTests.cs#L73-L100).

## Other Comments

I think snake case naming is pretty common, especially in the Ruby world, and should be supported out of the box. GitHub's API is probably the most popular example using this naming convention off the top of my head.

**Comment by hez2010** [comment]
@khellang I took your proposal contents to the top post, thanks!

**Comment by xsoheilalizadeh** [comment]
I've prepared a [benchmark ][1] that compares other implementation with #40003 PR.

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview8-013656
  [Host]     : .NET Core 3.0.0-preview8-28405-07 (CoreCLR 4.700.19.37902, CoreFX 4.700.19.40503), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview8-28405-07 (CoreCLR 4.700.19.37902, CoreFX 4.700.19.40503), 64bit RyuJIT

|                          Method |        Mean |      Error |     StdDev | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |------------:|-----------:|-----------:|-----:|-------:|------:|------:|----------:|
|  ToSnakeCaseStringBuilderBySpan |    85.97 ns |  1.0500 ns |  0.9821 ns |    1 | 0.0637 |     - |     - |     200 B |
| ToSnakeCaseNewtonsoftJsonBySpan |    86.35 ns |  1.7816 ns |  1.4877 ns |    1 | 0.0484 |     - |     - |     152 B |
|       ToSnakeCaseNewtonsoftJson |    88.89 ns |  0.8417 ns |  0.7461 ns |    2 | 0.0484 |     - |     - |     152 B |
|               ToSnakeCaseBySpan |   122.01 ns |  1.9066 ns |  1.6902 ns |    3 | 0.0253 |     - |     - |      80 B |
|                 ToSnakeCaseLinq |   365.75 ns |  4.4839 ns |  3.5008 ns |    4 | 0.1450 |     - |     - |     456 B |
|                ToSnakeCaseRegex | 2,108.71 ns | 40.6829 ns | 31.7625 ns |    5 | 
```
[1]:https://github.com/xsoheilalizadeh/SnakeCaseConversion/blob/master/SnakeCaseConversionBenchmark/SnakeCaseConventioneerBenchmark.cs

**Comment by xsoheilalizadeh** [comment]
I investigated a couple of weeks about this case and tried other approaches + #40003 that have better performance to having snake case naming policy. Finally, I got the best result and it have less allocation and fast execution. My last implementation is [ToSnakeCaseBySpan][1] . Shoud I open a new PR for this?
```
|                          Method |        Mean |      Error |     StdDev | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |------------:|-----------:|-----------:|-----:|-------:|------:|------:|----------:|
|               ToSnakeCaseBySpan |    27.47 ns |  0.4629 ns |  0.4330 ns |    1 | 0.0153 |     - |     - |      48 B |
|  ToSnakeCaseStringBuilderBySpan |    85.23 ns |  1.6495 ns |  1.3774 ns |    2 | 0.0637 |     - |     - |     200 B |
|       ToSnakeCaseNewtonsoftJson |    85.72 ns |  1.6418 ns |  1.4554 ns |    2 | 0.0484 |     - |     - |     152 B |
| ToSnakeCaseNewtonsoftJsonBySpan |    86.96 ns |  1.7060 ns |  1.5958 ns |    2 | 0.0484 |     - |     - |     152 B |
|                 ToSnakeCaseLinq |   353.42 ns |  3.9670 ns |  3.7108 ns |    3 | 0.1450 |     - |     - |     456 B |
|                ToSnakeCaseRegex | 2,056.69 ns | 29.5694 ns | 26.2125 ns |    4 | 0.1526 |     - |     - |     496 B |
```
[1]:https://github.com/xsoheilalizadeh/SnakeCaseConversion/blob/master/SnakeCaseConversionBenchmark/SnakeCaseConventioneerBenchmark.cs#L49

\cc @karelz

**Comment by hez2010** [comment]
@xsoheilalizadeh I think the name of a property is usually not too long, so maybe you can use stackalloc to reduce memory allocation on heap to reduce GC pressure?
```csharp
public unsafe ReadOnlySpan<char> ToSnakeCaseBySpan()
{
    ...
    Span<char> buffer = stackalloc char[bufferSize];
    ...
}
```

**Comment by xsoheilalizadeh** [comment]
Nice suggestion, but here is a question how I should return buffer, I've got following:
```
Cannot use local 'buffer' in this context because it may expose referenced variables outside of their declaration scope
```

**Comment by Gnbrkm41** [comment]
> Should I open a new PR for this?
Nope, not yet 😅 not until the API is approved.

**Comment by hez2010** [comment]
Because buffer now is allocated on the stack, you cannot return it directly since it will be disposed after return. To solve it, you can simply change `return buffer` to `return buffer.ToString()`;

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by karelz** [comment]
@ahsonkhan are we ready to review the API? @xsoheilalizadeh is eager to implement it ...

**Comment by xsoheilalizadeh** [comment]
I've figured out that the ToSnakeCaseBySpan can't pass these [tests][1], it's just compatible with pascal case properties, I've tried to make it compatible with other formats but unfortunately, I've failed and it'd be unreadable. I was hurrying to contribute and got a bad result, sorry.

[1]:https://github.com/hez2010/corefx/blob/c54ddc76da61c50576eff92e3add9caf7107f8af/src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs

**Comment by ahsonkhan** [comment]
This looks like a reasonable API addition. I was considering how common such formats are to warrant adding support built-in (rather than folks creating their own naming policy to support it), and looks like quite a few places use snake_case including Facebook, AWS, and Twitter APIs (aside from the fact that Json.NET supports it):
https://stackoverflow.com/questions/5543490/json-naming-convention

**Comment by terrajobst** [comment]
Makes sense. We also considered others (such as kebap casing) but it seems we should defer that until someone actually needs that.

**Comment by Gnbrkm41** [comment]
@xsoheilalizadeh, are you still working on it? :D you should be able to open PRs for it.

**Comment by xsoheilalizadeh** [comment]
Unfortunately no till I pass tests, I tried to pass tests but my implementation only support pascal-case naming for example `FirstName` will be `first_name` but in other cases like `IsCIA` it will be `Is_c_i_a` that isn't expected.

**Comment by hez2010** [comment]
@Gnbrkm41 How about the existing PR #40003

**Comment by Gnbrkm41** [comment]
I mean... sure, why not, as long as it has all the tests and it passes :D

**Comment by YohDeadfall** [comment]
@ahsonkhan Did the same thing already, so taking this.

**Comment by VassilAtanasov** [comment]
> Makes sense. We also considered others (such as kebap casing) but it seems we should defer that until someone actually needs that.

The lack of PascalCase option is a show stopper for our migration.
I believe there should be a "preserve case" option which will also work for us.

**Comment by khellang** [comment]
> The lack of PascalCase option is a show stopper for our migration.
> I believe there should be a "preserve case" option which will also work for us.

Isn't that the default? 🤔

https://github.com/dotnet/corefx/blob/296c0e71ddcc4885d30af18dea829c1923cfdc5c/src/System.Text.Json/src/System/Text/Json/Serialization/JsonDefaultNamingPolicy.cs#L9

**Comment by VassilAtanasov** [comment]
> > The lack of PascalCase option is a show stopper for our migration.
> > I believe there should be a "preserve case" option which will also work for us.
> 
> Isn't that the default? 🤔
> 
> https://github.com/dotnet/corefx/blob/296c0e71ddcc4885d30af18dea829c1923cfdc5c/src/System.Text.Json/src/System/Text/Json/Serialization/JsonDefaultNamingPolicy.cs#L9

Yes - that will work for us as the C# models use PascalCase properties

**Comment by Gnbrkm41** [comment]
the default is to use exact match IIRC. setting it to `null` also does that as well.

**Comment by niemyjski** [comment]
I'd prefer if we could just define our own naming strategies. We use our own naming method for snake_case because it occurred before Json.net came out with it and it's slightly different. So perhaps just allow us to set a naming policy or use the default one.

**Comment by khellang** [comment]
@niemyjski Of course you can. Just implement `JsonNamingPolicy` like the built-in ones and set that. Almost everything about the new serializer is configurable and/or extensible, these issues are usually just about including more stuff in the box.

**Comment by sveinfid** [comment]
Please could KebabCase also be added?

It should be a *very* minor addition - basically the same as SnakeCase but with `-` in place of `_`

E.g. requests sent to Apple Push Notification Service **require** kebab-case JSON fields: https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification#2943360

**Comment by ahsonkhan** [comment]
@sveinfid - There was precedence for snake casing (e.g. in newtonsoft json) and enough user need to provide built-in support for it. I am not sure that's the case for KebabCase. P**lease file a separate issue (with the API proposal) and we can consider it**, but my hunch is, this is something that the caller would have to implement (like any other custom casing/naming strategy) and doesn't seem common enough to require built-in support.

Like you said, the impl could be relatively straight forward (take existing snake case impl and replace the delimiter), which means there is also little value in providing one built-in.

**Comment by layomia** [comment]
From @YohDeadfall in https://github.com/dotnet/runtime/issues/673:

> Originally proposed by @hez2010 in [dotnet/corefx#39564](https://github.com/dotnet/corefx/issues/39564).
> 
> @ahsonkhan, could you add the original issue here and important comments from [dotnet/corefx#41354](https://github.com/dotnet/corefx/pull/41354)?
> 
> @karelz We need your opinion here, since @khellang asked a question about altering the existing behavior of the camel case policy too follow common translator rules for the JS world as splitting a string by words and concatenating it back using some separator and applying a policy to change casing for each word. It's a breaking change and @stephentoub likes it. Me too. See [dotnet/corefx#41354 (comment)](https://github.com/dotnet/corefx/pull/41354#issuecomment-554958476).
> 
> I still would like to drive this feature.
> 
> @steveharter @layomia
